rfcs/property-getter-setter.md

```diff
@@ -0,0 +1,163 @@
+particular, this means there are read-only arrays `{get T}`, that are
+covariant, so we have a solution to the "covariant array problem":
+
+```lua
+local dogs: {Dog}
+function f(a: {get Animal}) ... end
+f(dogs)
+```
+
+It is sound to allow this program, since `f` only needs read access to
+the array, and `{Dog}` is a subtype of `{get Dog}`, which is a subtype
+of `{get Animal}`.  This would not be sound if `f` had write access,
+for example `function f(a: {Animal}) a[1] = Cat.new() end`.
+
+### Functions
+
+Functions are not normally mutated after they are initialized, so
+```lua
+function t.f() ... end
+function t:m() ... end
+```
+should introduce a getter, but not a setter. If developers want a mutable function,
+they can use the anonymous function version
+```lua
+t.f = function() ... end
+```
+
+Methods in classes should be read-only by default.
+
+*This is a possibly breaking change.*
+
+### Classes
+
+Getter and setter types apply to classes as well as tables.
+
+Many of the Roblox APIs an be marked as having getters but not
+setters, which will improve accuracy of type checking for Roblox APIs.
+
+### Metatables
+
+Metatables should be read-only by default.
+
+*This is a possibly breaking change.*
+
+### Require
+
+A `require`d module's properties should be read-only by default.
+
+*This is a possibly breaking change.*
+
+### Why separate the getter from the setter?
+
+Separate getters and setters were introduced to TypeScript in response
+to use-cases of APIs which performed data conversion using getter and
+setter methods (the equivalent of Lua `__index` and `__newindex`
+metamethods).
+
+As well as those use-cases, separating read and write access,
+considerably simplifies variance, which is an important part of type inference:
+
+* Read-only properties are *covariant*: if `T` is a subtype of `U`
+  then `{ get p: T }` is a subtype of `{ get p: U }.
+* Write-only properties are *contravariant*: if `T` is a supertype of `U`
+  then `{ set p: T }` is a subtype of `{ set p: U }.
+
+This allows type inference to provide a principal type to tables.
+For example in:
+
+```lua
+local x = { p: nil }
+function g(d : Dog?) x.p = d end
+function h(): Animal? return x.p end
+```
+
+it is not obvious what type to infer for `x`, should it be `{ p: Dog? }` or `{ p: Animal? }`?
+If there are separate getter and setter types, there is a most general type, which is
+`{ get p: Animal?, set p: Dog? }`.
+
+With separate getter and setter types, every position in a type is either
+covariant or contravariant, and *there are no more uses of invariance*.
+This simplifies the implementation (all our special-casing for
+invariance is no longer needed) and makes possible techniques such as local
+type inference (which relies on variance to infer a principal type).
+
+### Related work
+
+This proposal is based on
+[TypeScript 4.3](https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#separate-write-types).
+getters and setters.
```

Comment:
I guess it's not clear to me why we should solve this problem with separate types instead of with access control. The alternatives section mentions this in passing but I'm not sure I understand the rationale. The rationale for not introducing separate types is that it makes it easier to wrap your head around the complete definition of the table type, and we should strongly prefer simpler mental models to more complicated ones.

Reply:
Well the question is what to do in cases which typecheck with separate getters/setters, is it better to produce a type error even though it's a false positive, or infer a more complex type? It's not obvious, either way we're adding complexity to the user's life (either by trying to understand a horrible type error, or a complex inferred type).

Reply:
My point is, I feel like this RFC doesn't really explore ways for us to not split the types, instead just focusing on that as the right solution. It's not clear to me that the type error is necessarily "horrible" without interesting examples.

I'm also compelled to say that TypeScript can serve as a useful inspiration but it's not clear to me if it's a good model to follow in this case. In particular, TypeScript allows properties to have custom get & set code, so the motivation there is clear - allow setters to have a different, broader, type, so that the setter can convert it to a narrower storage type.

For us, table types aren't just types of access, they are also types of storage in that annotating a field as T says "this field only ever stores values of type T". This proposal suggests switching to types of access and I'm not sure that's the right tradeoff.

Reply:
I added some more text around the alternative design. I can spell it out in more detail if you like, though this will be wasted effort if we decide to go with separate getter/setter types.

TypeScript is not the only place splitting getter/setter types, though it's probably the most widely-used language that does this. I think it goes back to Reynolds and Idealized Algol, though I first came across it in the context of Pict sometime in the 1990s.